sympy/solvers/solveset.py

```diff
@@ -1754,6 +1763,33 @@ def _is_logarithmic(f, symbol):
     return rv
 
 
+def _solve_as_lambert(lhs, rhs, symbol, domain):
+    r"""
+    Computes the equations whose solutions will be given in terms
+    of the `LambertW` function. Returns `None` if it fails doing so.
+
+    Example
+    =======
+
+    >>> from sympy.solvers.solveset import _solve_as_lambert
+    >>> from sympy import symbols, exp, S
+    >>> x = symbols('x')
+    >>> _solve_as_lambert(x*exp(x), 1, x, S.Reals)
+    FiniteSet(LambertW(1))
+
+    """
+    poly = lhs.as_poly()
+    gens = _filtered_gens(poly, symbol)
+    try:
+        soln = _solve_lambert(lhs - rhs, symbol, gens)
+    except NotImplementedError:
+        return None
+    else:
+        if domain.is_subset(S.Reals) and len((lhs - rhs).atoms(Symbol)) == 1:
+            soln = [s for s in soln if s.is_real]
+        return FiniteSet(*soln)
```

Comment:
Why does this check `len((lhs - rhs).atoms(Symbol)) == 1`?

Reply:
This checks for no. of symbols.
If it's more than 1 there are more symbols(the solution may contain a `symbol`) and if it's equal to 1 it means the solution will be having no symbol.

Reply:
But why should that affect whether or not we want to remove the solutions that are not real?

Actually I think that this check should be removed and it should just be:
```
return FiniteSet(*soln) & domain
```

Reply:
> Actually I think that this check should be removed and it should just be:
> 
> return FiniteSet(*soln) & domain

Actually this is logic already under implementation. But it's necessary to have a check since without that check it doesn't give the desired results as of now.

> But why should that affect whether or not we want to remove the solutions that are not real?

Firstly, this should be understood that having a symbol in solution and not having a solution are different cases. 

If there is a symbol in solution then it could have `I` since other symbol may have the value that may adjust finally to a real value and for the same reason `Intersection` set is used  .

If there Is no symbol then there should be definitely a check for weather it's a real or not. Hence the reason for a check (i.e. `len((lhs - rhs).atoms(Symbol)) == 1:` )  is compulsory.

Reply:
I hope you've got the reason for the check (`len((lhs - rhs).atoms(Symbol)) == 1:`).

Reply:
With check (`if ... len((lhs - rhs).atoms(Symbol)) == 1:`).
(no symbol in output or `len((lhs - rhs).atoms(Symbol)) == 1`)
```
assert solveset_real(3*x + 5 + 2**(-5*x + 3), x) == S.EmptySet

    assert solveset_real(2*x + 5 + log(3*x - 2), x) == \
        FiniteSet(Rational(2, 3) + LambertW(2*exp(Rational(-19, 3))/3)/2)
```
Without check

```
>>> solveset_real(3*x + 5 + 2**(-5*x + 3), x)
Intersection(FiniteSet(-5/3 + LambertW(-5*(-1024*2**(1/3) - 1024*2**(1/3)*sqrt(3)*I)*log(2)/3)/(5*log(2)), -5/3 + LambertW(-5*(-1024*2**(1/3) + 1024*2**(1/3)*sqrt(3)*I)*log(2)/3)/(5*log(2))), Reals)

or

    ⎧       ⎛   ⎛       3 ___        3 ___     ⎞        ⎞         ⎛   ⎛       3 ___        3 ___     ⎞        ⎞⎫
    ⎪       ⎜-5⋅⎝- 1024⋅╲╱ 2  - 1024⋅╲╱ 2 ⋅√3⋅ⅈ⎠⋅log(2) ⎟         ⎜-5⋅⎝- 1024⋅╲╱ 2  + 1024⋅╲╱ 2 ⋅√3⋅ⅈ⎠⋅log(2) ⎟⎪
    ⎪      W⎜───────────────────────────────────────────⎟        W⎜───────────────────────────────────────────⎟⎪
    ⎨  5    ⎝                     3                     ⎠    5    ⎝                     3                     ⎠⎬
ℝ ∩ ⎪- ─ + ──────────────────────────────────────────────, - ─ + ──────────────────────────────────────────────⎪
    ⎪  3                      5⋅log(2)                       3                      5⋅log(2)                   ⎪
    ⎩                                                                                                          ⎭



>>> solveset_real(2*x + 5 + log(3*x - 2), x) 
Intersection(FiniteSet(LambertW(2*exp(-19/3)/3)/2 + 2/3, 2/3 + LambertW(-exp(-19/3)/3 - sqrt(3)*I*exp(-19/3)/3)/2, 2/3 + LambertW(-exp(-19/3)/3 + sqrt(3)*I*exp(-19/3)/3)/2), Reals)

or

    ⎧     ⎛   -19/3         -19/3⎞       ⎛   -19/3         -19/3⎞   ⎛   -19/3⎞    ⎫
    ⎪     ⎜  ℯ        √3⋅ⅈ⋅ℯ     ⎟       ⎜  ℯ        √3⋅ⅈ⋅ℯ     ⎟   ⎜2⋅ℯ     ⎟    ⎪
    ⎪    W⎜- ────── - ───────────⎟      W⎜- ────── + ───────────⎟  W⎜────────⎟    ⎪
    ⎨2    ⎝    3           3     ⎠  2    ⎝    3           3     ⎠   ⎝   3    ⎠   2⎬
ℝ ∩ ⎪─ + ─────────────────────────, ─ + ─────────────────────────, ─────────── + ─⎪
    ⎪3               2              3               2                   2        3⎪
    ⎩                                                                             ⎭
```

Reply:
I still don't understand why the check makes a difference.

Reply:
The check here try to see no of symbols ( the check is ` len((lhs - rhs).atoms(Symbol)) == 1:`).

If there is more than one symbol that implies that the result definitely consists of at least one symbol (i.e. `total no of symbols -1 ` ). 
So for such case, there should be a need for the check!

If the expression contains only one symbol this implies that the `result` will be either a `ConditionSet` or a result without any symbol.

Reply:
e.g. 
Here in the below cases, the expression consists of only one symbol ie `x` so the output doesn't consist of any symbol so there is a need for the check so that the output should not contain any nonreal values
```
solveset_real(x*exp(x) - 1, x) == FiniteSet(LambertW(1))

solveset_real((a/x + exp(x/2)).diff(x), x) == Complement(Intersection(\
        FiniteSet(4*LambertW(-sqrt(2)*sqrt(a)/4), 4*LambertW(sqrt(2)*sqrt(a)/4)), S.Reals), FiniteSet(0))
```
Now considering the cases where there is more than one symbol i.e. (`x`,`p` and `x`,`a`) here there is the possibility that `p` or `a` may have such a value that it may give a real solution thus there is no need for check for these cases. Instead, we use `Intersection` so that it will be more precise.
```
p = symbols('p', positive=True)
 solveset_real(3*log(p**(3*x + 5)) + p**(3*x + 5), x) == \
        Intersection(FiniteSet(Rational(-5, 3) - LambertW(Rational(1, 3))/(3*log(p)),
         Rational(-5, 3) - LambertW(Rational(-1, 6) - sqrt(3)*I/6)/(3*log(p)),
          Rational(-5, 3) - LambertW(Rational(-1, 6) + sqrt(3)*I/6)/(3*log(p))), S.Reals)

solveset_real((a/x + exp(x/2)).diff(x, 2), x) == Complement(Intersection(\
        FiniteSet(6*LambertW((-a)**(Rational(1, 3))/3), \
            6*LambertW(-(-a)**(Rational(1, 3))/6 - sqrt(3)*I*(-a)**(Rational(1, 3))/6), \
                6*LambertW(-(-a)**(Rational(1, 3))/6 + sqrt(3)*I*(-a)**(Rational(1, 3))/6)), S.Reals), FiniteSet(0))
```

Reply:
I hope this is more clear.

Reply:
Does the `Intersection` not automatically check `is_real` anyway?
```python
In [14]: FiniteSet(1, 2)
Out[14]: {1, 2}

In [15]: FiniteSet(1, 2) & Reals
Out[15]: {1, 2}

In [16]: FiniteSet(I, 2*I) & Reals
Out[16]: ∅

In [17]: x = Symbol('x', real=True)

In [18]: FiniteSet(x, 2*x) & Reals
Out[18]: {x, 2⋅x}
```

Reply:
I've tried this way also but this is not working here at all : (

Reply:
```
>>> b = Intersection(FiniteSet(LambertW(2*exp(-19/3)/3)/2 + 2/3, 2/3 + LambertW(-exp(-19/3)/3 - sqrt(3)*I*exp(-19/3)/3)/2, 2/3 + LambertW(-exp(-19/3)/3 + sqrt(3)*I*exp(-19/3)/3)/2), Reals)
>>> b & Reals
Intersection(FiniteSet(0.667258001414628, 0.666666666666667 + LambertW(-0.000592034515244793 - 0.000592034515244793*sqrt(3)*I)/2, 0.666666666666667 + LambertW(-0.000592034515244793 + 0.000592034515244793*sqrt(3)*I)/2), Reals)
```

Reply:
Oh, I see why this code makes a difference: it removes solutions that have `is_real=None`. That's not generally valid though because `is_real=None` means "don't know" rather than "not real". I think it's better to return an intersection in this case. Possibly Lambert.is_real can be improved though.

Reply:
What about 

```
if domain.is_subset(S.Reals) and len((lhs - rhs).atoms(Symbol)) == 1:
            soln = [s for s in soln if not (s.n()).has(I)]
        return FiniteSet(*soln) 
```
IMHO, I think this could be a lot better.

Reply:
All of the suggestions you have made for removing the non-real solutions will also fail in some cases by removing a solution that actually is real. I think it is better just to return the `Intersection` if `is_real` can not determine that the expression is not real.